### PR TITLE
ZIO with contextual effects

### PR DESCRIPTION
**Pull request by wi101**

### CODE REVIEW

File: project/ScalazBuild.scala [hunk 1]
```diff
@@ -25,8 +25,7 @@ object Scalaz {
     "-Xlint:_,-type-parameter-shadow",
     "-Xsource:2.13",
     "-Ywarn-numeric-widen",
-    "-Ywarn-value-discard",
-    "-Xfatal-warnings"
```
Comment (NeQuissimus) [inline project/ScalazBuild.scala]: Why are you dropping fatal warnings?

Comment (wi101) [reply project/ScalazBuild.scala]: There were some warnings, thanks, I added them now

**Comment by NeQuissimus** [comment]
you guys... breaking everything I've built so far and not making me feel bad about it...

**Comment by tabdulradi** [comment]
I asked the following question in the meetup tonight
> Is it possible to partially provide environment? Like `R1 => ZIO[R1 with R2, E, A] => ZIO[R2, E, A]` 

Now thinking about using HLists instead of intersection types. `R1 => ZIO[R1 :: R2 :: HNil, E, A] => ZIO[R1 :: HNil, E, A]` should be possible, what do you think? 🤔

File: core/shared/src/main/scala/scalaz/zio/ZIO.scala [hunk 2]
```diff
@@ -1,18 +1,1604 @@
...
+    repeatOrElseEither[R1, B, E2, B](schedule, orElse).map(_.merge)
+
+  /**
+   * Repeats this effect with the specified schedule until the schedule
+   * completes, or until the first failure. In the event of failure the progress
+   * to date, together with the error, will be passed to the specified handler.
+   */
+  final def repeatOrElseEither[R1 <: R, B, E2, C](
+    schedule: Schedule[R1, A, B],
+    orElse: (E, Option[B]) => ZIO[R1 with Clock, E2, C]
```
Comment (NeQuissimus) [inline core/shared/src/main/scala/scalaz/zio/ZIO.scala]: Why does `R1` need the `with Clock` here but not in the `orElse` above?
I am clearly missing something :)

File: core/shared/src/main/scala/scalaz/zio/ZIO.scala [hunk 3]
```diff
@@ -1,18 +1,1604 @@
...
+   * Returns information about the current fiber, such as its fiber identity.
+   */
+  final def descriptor: UIO[Fiber.Descriptor] = ZIO.Descriptor
 }
-object ZIO extends Serializable {
-  final def apply[F[_, _]](implicit F: ZIO[F]): ZIO[F] = F
 
-  implicit final val ZIOIO: ZIO[IO] = new ZIO[IO] {
-    final def liftZIO[E, A](io: IO[E, A]): IO[E, A] = io
+trait ZIO_E_Any extends ZIO_E_Throwable {
```
Comment (NeQuissimus) [inline core/shared/src/main/scala/scalaz/zio/ZIO.scala]: I am not a fan of these strange trait names...

**Comment by NeQuissimus** [comment]
It would be great to provide additional documentation for the microsite.
`ZIO` is no longer immediately "familiar" and I feel the improvements in this PR require additional information. The standard `IO[A]` most people are familiar with. `IO[E, A]` was easily explained. The concept of the environment here is not as intuitive.

File: core/shared/src/main/scala/scalaz/zio/ZIO.scala (hunk 2 above)
Comment (wi101) [reply core/shared/src/main/scala/scalaz/zio/ZIO.scala]: in `repeatOrElseEither` we used Schedule which has `Clock` now

**Comment by dwijnand** [comment]
Additionally, I fear this introduces the ambiguity of when one says "zio"/"ZIO" they mean the library or the datatype in that library.  I caught myself today saying "scalaz-zio's ZIO datatype" which is a mouthful.  Maybe I should've said "scalaz's ZIO datatype", which is defined in the scalaz-zio foundation library.

File: core/shared/src/main/scala/scalaz/zio/random/Random.scala [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    val nextGaussian: ZIO[R, Nothing, Double]
+    def nextInt(n: Int): ZIO[R, Nothing, Int]
+    val nextInt: ZIO[R, Nothing, Int]
+    val nextLong: ZIO[R, Nothing, Long]
+    val nextPrintableChar: ZIO[R, Nothing, Char]
+    def nextString(length: Int): ZIO[R, Nothing, String]
+  }
+  trait Live extends Random {
+    val random: Service[Any] = new Service[Any] {
+      import scala.util.{ Random => SRandom }
```
Comment (jmcardon) [inline core/shared/src/main/scala/scalaz/zio/random/Random.scala]: Excuse my intrusion on this package given my absence in scala, but this is effectively just using a global random instance, even if it's observably pure. I'm not precisely sure that's a great thing to expose.

File: project/ScalazBuild.scala (hunk 1 above)
Comment (wi101) [reply project/ScalazBuild.scala]: I don't know how my comment disappear :D !

Comment (wi101) [reply project/ScalazBuild.scala]: @NeQuissimus  I added this plugin: 
```sbt

addSbtPlugin("com.thoughtworks.sbt-api-mappings" % "sbt-api-mappings"  % "3.0.0")
```

File: microsite/src/main/tut/interop/catseffect.md [hunk 5]
```diff
@@ -15,7 +15,7 @@ Checkout `interop-cats` module for inter-operation support.
 For convenience we have defined an alias as follow:
 
 ```scala
-  type Task[A] = IO[Throwable, A]
+  type Task[A] = Task[A]
```
Comment (wongelz) [inline microsite/src/main/tut/interop/catseffect.md]: `type Task[A] = Task[A]`?

**Comment by emilypi** [comment]
Killing it 👍 

I'll give a review this weekend.

**Comment by landerlo** [comment]
> I asked the following question in the meetup tonight
> 
> > Is it possible to partially provide environment? Like `R1 => ZIO[R1 with R2, E, A] => ZIO[R2, E, A]`
> 
> Now thinking about using HLists instead of intersection types. `R1 => ZIO[R1 :: R2 :: HNil, E, A] => ZIO[R1 :: HNil, E, A]` should be possible, what do you think? thinking

@tabdulradi We would lose type inference and incur a severe dispatch performance penalty. What you want should be trivial in dotty when we have proper intersection types.

**Comment by jdegoes** [comment]
@tabdulradi Should be possible! 💡 ZIO itself makes no requirements on environmental subtyping, but subtyping is supported if you want to use it (via auto-widening in flatMap).

Also some really interesting things become possible with higher-kinded `R`. 

This is an area of active exploration in Scala... 🔎

File: core/shared/src/main/scala/scalaz/zio/ZIO.scala (hunk 3 above)
Comment (jdegoes) [reply core/shared/src/main/scala/scalaz/zio/ZIO.scala]: I'll clean them up in followup PR.

**Comment by jdegoes** [comment]
There's some good feedback here. Since this has already had 3 full reviews prior to opening and several more since then, I'm going to go ahead and merge, and address minor cleanup and major documentation in followup PRs.

**Comment by oscarvarto** [comment]
@wi101 and @jdegoes Thanks for all your efforts.
I'm very eager to read your documentation.

**Comment by edmundnoble** [comment]
@tabdulradi That will likely require macros. I've seen people try to construct `A with B` given an `A` and a `B`, but never seen success.

File: core/shared/src/main/scala/scalaz/zio/Managed.scala [hunk 6]
```diff
@@ -73,49 +89,49 @@ final case class Managed[+E, +R](reserve: IO[E, Managed.Reservation[E, R]]) { se
...
 
 object Managed {
-  final case class Reservation[+E, +A](acquire: IO[E, A], release: IO[Nothing, _])
+  final case class Reservation[-R, +E, +A](acquire: ZIO[R, E, A], release: UIO[_])
 
   /**
    * Lifts an `IO[E, R]`` into `Managed[E, R]`` with a release action.
    */
-  final def make[E, R](acquire: IO[E, R])(release: R => IO[Nothing, _]): Managed[E, R] =
+  final def make[R, E, A](acquire: ZIO[R, E, A])(release: A => UIO[_]): Managed[R, E, A] =
```
Comment (ghostdogpr) [inline core/shared/src/main/scala/scalaz/zio/Managed.scala]: Shouldn't `release` be a `ZIO[R1, Nothing, _]`? Similar to `IO.bracket`?

**Pull request merged**